### PR TITLE
#12985: tweak functor error msgs by prefering changes on the left

### LINKED ISSUE #12985: Improve type error explanation for partially applied functor
I found the error message for partially applied functors to be confusing when the argument doesn't type check:

```ocaml
module type A = sig type a val missing : a end
module type B = sig type b end

module Make (A : A) (B : B) = struct end

module A_impl = struct type a = unit end

module Partial = Make (A_impl)
(*               ^^^^^^^^^^^^^
Error: The functor application is ill-typed.
       These arguments:
         A_impl
       do not match these parameters:
         functor (A : A) (B : B) -> ...
       1. An argument appears to be missing with module type A
       2. Modules do not match:
            A_impl : sig type a = unit end
          is not included in
            B
*)
```

It looks like some magic is happening to guess which functor parameter is missing, but when there are no good choices it introduces the missing arguments on the left rather than the right, leading to an unexpected error message (... assuming the partial application was otherwise correct)

```ocaml
module Partial = Make (<missing> : A) (A_impl : B)  (* current error *)

(* I expected the error to keep [A_impl] as the first parameter,
   optionally mentioning that the second argument [B] is missing *)
module Partial = Make (A_impl)
(*               ^^^^^^^^^^^^^
Error: The functor application is ill-typed.
       These arguments:
         A_impl
       do not match these parameters:
         functor (A : A) (B : B) -> ...
       1. Modules do not match:
            A_impl : sig type a = unit end
          is not included in
            A
          The value `missing' is required but not provided
       2. An argument appears to be missing with module type B
*)
```

(Obviously low priority considering the rarity of partially applied functors!)

### PR DESCRIPTION
**Pull request by Octachron**
This PR improves the error message for partially applied functor by making the diffing algorithm prefers paths with optimal weight that end with deletions or additions.

This one-line change is sufficient to switch the error message for  
```ocaml
module type x = sig type x end
module type y = sig type y end
module X = struct type x end
module Y = struct type y end
module W = struct type w end
module F(A:x)(B:y)(C:x) = struct end
module M = F(W)
```
from
```
module M = F(W);;
Error: The functor application is ill-typed.
       These arguments:
         W
       do not match these parameters:
         functor (A : x) (B : y) (C : x) -> ...
       1. An argument appears to be missing with module type x
       2. An argument appears to be missing with module type y
       3. Modules do not match: W : sig type w = W.w end is not included in x
```
to
```
Error: Modules do not match: sig type w = W.w end is not included in 
       x
     The type x is required but not provided
```

close #12985

### CODE REVIEW

File: testsuite/tests/typing-misc/records.ml [hunk 1]
```diff
@@ -268,7 +268,7 @@ Line 1, characters 0-30:
 1 | type perm = d = {y:int; x:int}
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 Error: This variant or record definition does not match that of type "d"
-       Fields "x" and "y" have been swapped.
+       Field "y" has been moved from position 1 to 2.
```
Comment (gasche) [inline testsuite/tests/typing-misc/records.ml]: I think that this is a regression.

File: testsuite/tests/typing-modules/functors.ml [hunk 2]
```diff
@@ -2054,9 +2104,21 @@ Error: The functor application "Set.Make(Set)(A)" is ill-typed.
          Set A
        do not match these parameters:
          functor (Ord : Set.OrderedType) -> ...
-       1. The following extra argument is provided Set : (module Set)
-       2. Modules do not match:
-            A : sig type a = A.a end
+       1. Modules do not match:
+            Set : (module Set)
+          is not included in
+            Set.OrderedType
```
Comment (gasche) [inline testsuite/tests/typing-modules/functors.ml:2110]: My impression is that this error message is also worse than before -- this would be clearer if `A` had type `OrderedType`, which is how I tend to read this example.

File: testsuite/tests/typing-modules/functors.ml [hunk 3]
```diff
@@ -1667,16 +1712,21 @@ Error: The functor application "Bar(A)(FiveArgsExt)(TY)(TY)(TY)(TY)(TY)" is ill-
...
        1. Module A matches the expected module type A
-       2. An argument appears to be missing with module type Ext
-       3. An argument appears to be missing with module type B
-       4. Module FiveArgsExt matches the expected module type Ext
-       5. Module TY matches the expected module type ty
-       6. Module TY matches the expected module type ty
-       7. Module TY matches the expected module type ty
-       8. Module TY matches the expected module type ty
-       9. Module TY matches the expected module type ty
+       2. Module FiveArgsExt matches the expected module type Ext
```
Comment (gasche) [inline testsuite/tests/typing-modules/functors.ml:1717]: This is (slightly) better than before.

File: testsuite/tests/typing-modules/functors.ml [hunk 4]
```diff
@@ -1538,9 +1586,9 @@ Error: Signature mismatch:
             sig end
           The type "wrong" is required but not provided
        2. Module types $S2 and $T2 match
-       3. An extra argument is provided of module type X.T
+       3. Module types X.T and X.T match
        4. Module types X.T and X.T match
-       5. Module types X.T and X.T match
+       5. An extra argument is provided of module type X.T
```
Comment (gasche) [inline testsuite/tests/typing-modules/functors.ml:1591]: This is slightly better than before.

File: testsuite/tests/typing-modules/functors.ml [hunk 5]
```diff
@@ -383,8 +428,8 @@ Error: This application of the functor "F" is ill-typed.
        do not match these parameters:
          functor (A : ...) (B : ...) (C : ...) -> ...
        1. Module $S1 matches the expected module type
-       2. The following extra argument is provided X : sig type x = int end
-       3. Module X matches the expected module type
+       2. Module X matches the expected module type
+       3. The following extra argument is provided X : sig type x = int end
```
Comment (gasche) [inline testsuite/tests/typing-modules/functors.ml:432]: These look almost identical but I agree that the new order is slightly better.

**Review by gasche — COMMENTED** [review]
Looking at the testsuite examples, my impression is that the change introduces some regressions as well as some improvements.

File: testsuite/tests/typing-modules/functors.ml (hunk 2 above)
Comment (Octachron) [reply testsuite/tests/typing-modules/functors.ml:2110]: This case is completely symmetrical? We are going from a `Delete-Change` to `Change-Delete`? If `A` was matching, the path that will be selected would be `Delete-Match` .

Comment (gasche) [reply testsuite/tests/typing-modules/functors.ml:2110]: To me as a human, it is very clear that `Set` cannot possibly be an OrderedType choice, while A looks like a type that may be intended to be an OrderedType (but maybe `compare` is mis-spelled or something?).

(The actual error may be that `Set` was written when say `ISet` was meant, which happens to be an ordered type. But then showing in details the signature of `Set` is *also* irrelevant to that sort of error, so the new message is also mostly noise in this case.)

File: testsuite/tests/typing-misc/records.ml (hunk 1 above)
Comment (Octachron) [reply testsuite/tests/typing-misc/records.ml]: This was a corner case because the existing weight considered that a move of one field 
```
a   b   a
↓-------↑
```
(aka a delete + insertion ) was a good as an exchange of two fields
```
a  b
↓  ↓
b  a
```
(aka two change) I have fixed this case by  tweaking the weights to prefer a swap.

**Comment by gasche** [comment]
Preferring swaps to moves is very reasonable -- as a change of its own I would be happy to approve it. On the other hand, I don't understand what this second commit does in the details, it throws away a `t.types_match` that may have been a good idea. (Maybe some comment on the intent between the weights would be nice.)

File: testsuite/tests/typing-modules/functors.ml (hunk 2 above)
Comment (art-w) [reply testsuite/tests/typing-modules/functors.ml:2110]: The module `A` has type `sig type a end` so it's not clear to me how it can be considered more similar to `OrderedType` than `Set` without requiring human-level sensitivity (although it generated a smaller error message when picked, but I don't know why the missing type `t` and `compare` value were not listed before)

Comment (gasche) [reply testsuite/tests/typing-modules/functors.ml:2110]: I agree. I'm not saying that whatever criterion we use should recognize one from the other, just that the new error message is (for humans) slightly worse than before.

**Comment by gasche** [comment]
After thinking more about this, I have mellowed on this PR. I think that there is probably a reasonable idea behind it, it is not just clear from looking at the diff, and that the regression in the existing messages are probably taken into account reasonably well by your new, additional change.

I remain disconcerted from the fact that I have no clue what is going on when looking at the diff. Could you:

1. Explain how the code changes realize the high-level specifications that you describe, and
2. Ensure that there are enough comments in the source, so that code readers know what the high-level specification is (and maybe how it is realized by the code)?

**Comment by Octachron** [comment]
What happens is that the line
```ocaml
    select_best_proposition [del;insert;diag]
```
enforces an implicit reverse lexicographic order on the minimal weight paths.

In more details, consider the case where the three candidate have the same weight, we end-up with three paths ending with `... Del`, `... Insert`, `... Change`.  The code then selects as a representative of the weight-equivalence class the first one.
This is equivalent to choosing the smallest best-weight path according to the order ` Del < Insert < Change` for the last part of the path.

By induction, since we have applied this choice for the all preceding matrix position, this means that the selected path among the paths with minimal weight is the smallest one according to the corresponding reverse lexicographic order.

Previously, we (I?) choose the order `Change < Del < Insert`, with the idea that all those paths were equivalent.

However, for the sake of partial functor application, we already have in place a elision mechanism for the `Insert` suffix of patches. This was done because we always compare the full list of parameters to list of concrete arguments, and thus
```ocaml
module F: A -> B -> C - > ...
module M = F(A)
```
results in a `Match → Insert → Insert`  patch where the `Insert` suffix is non-informative. 

Unfortunately, this elision mechanism was triggering less often that it should have because of our choice of order for optimal path.

For instance, if we apply the previous functor to an argument that doesn't match any parameter
```ocaml
module F: A -> B -> C -> ...
module M = F(D)
```
the paths with optimal weights are:
*  `Change → Insert → Insert`
*  `Insert → Change → Insert`
*  `Insert → Insert → Change`

 And with the `Change< Insert < Del` order, we selected the last one completely disabling the `Insert` suffix elision mechanism. The first change in this `PR` avoids this situation by enforcing that `Insert<Change`. Not that the we also changed the relation of `Del`; but that is not necessary and we could choose either
 
 * `Insert < Change < Del`
 * `Insert < Del < Change`
 * `Del < Insert < Change` (which the one used in this PR).

**Comment by gasche** [comment]
I understand the explanation, thanks. Two questions:

1. Is there a symmetric situation with `Change` and `Del`, when a functor is over-applied, that would justify not using `Change < Del`?
2. Could you do this differently (but I'm not suggesting that you must, I'm happy with the minimal change here if properly explained) by generating a different `Insert_at_end` (and symmetrically `Delete_at_end` if relevant) action that is only emitted when diffing against the empty sequence, and would be preferred over `Insert` (and `Delete`)?

**Comment by Octachron** [comment]
1. For over-applied functors, it doesn't feel accurate to consider that arguments on the right are more likely to be erroneous. For deletions, it is the asymmetry of partial applications that breaks the position-invariance of error likelihood for `Deletion`, I don't see a similar phenomenon for insertion.
2. Abstract module types makes this less easy than it appear because they make the shape of the diffing matrix dynamic, and thus we cannot identify the last rows (or columns) before we reach the end corner.
Typically, with
```ocaml
module F(X:sig module type T module M:T end) = X.M
module type A = sig type a end
module B = struct type b end
module Three = struct
  module M(_:A)(_:A)(_:A) = struct end
  module type T = module type of M
end
module E = F(Three)(B)
```
we start with a 2x1 matrix and end with a 2x4 matrix. Selecting the path representative using the reverse lexicographic order bypasses the issue.
```
Error: This application of the functor F is ill-typed.
       These arguments:
         Three B
       do not match these parameters:
         functor (X : ...) A -> ...
       1. Module Three matches the expected module type
       2. Modules do not match:
            B : sig type b = B.b end
          is not included in
            A
          The type a is required but not provided
```

**Comment by gasche** [comment]
Status: I understand the explanation for the first change. You haven't tried explaining the second change, and/or reflecting the explanations in the codebase.

**Comment by Octachron** [comment]
I have added the explanation for the path preference in the diffing matrix in the code.

I have also tweaked a bit further the weights for the definition diffing, with an explanation of the constraints involved:

- Zeroth, without loss of generality we can choose `w Delete` = `w Insert`.
- First, we want `w Change < w Insert + w Delete`, otherwise `Change` patch element are not useful.
- Second, if we want to prefer consecutive swaps to moves, this implies that 
       `w Swap < w Move ` ⇒ ` w (Change^2) < w (Insert Delete)` ⇒ `w Change < w Delete`
- Third, we prefer bracketed `Delete ... Insert` pairs that allow more matches to long sequence of `Change`s. 
   In other words, for some D ∈ ℕ* and a large enough  R ∈ ℕ*, we should have
     `w (Delete^D Keep^R Insert^D) < w (Change^(D+R))` ⇒  `w Change > (2 D)/(D+R) * (w Delete)`.
  Note that the case `D=1,R=1` is incompatible with the inequation `w Change<w Delete` above. 
  If we choose `R = D + 1` for `D<5`, we can specialize the  inequation to ` w Change > (10 / 11) * w Delete `.
- Fourth, the ordering for `Type_only_change`, `Name_only_change` and `Name_and_type_change` determine which change is considered as more likely mistake, `w Type_only_change < w Name_only_change < w Name_and_type_change` seems like a good ordering.

**Review by gasche — APPROVED** [review]
Great! These are more explanations than anyone ever knew we could have, so this is a definite improvement and I am happy to Approve.

**Comment by gasche** [comment]
check-typo would like a Changes entry
